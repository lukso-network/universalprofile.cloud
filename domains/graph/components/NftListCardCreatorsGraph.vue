<script setup lang="ts">
type Props = {
  asset: Asset
}

export type VerifyStatus = 'verified' | 'unverified' | 'partial'

const CREATOR_SHOW_LIMIT = 4
const props = defineProps<Props>()

const asset = computed(() => props.asset)
const assetAddress = computed(() => asset.value?.address)
const creators = computed(() => {
  let items = props.asset?.tokenCreatorsData || []

  if (items.length === 0 && props.asset?.ownerData) {
    items = [asset.value?.ownerData as Creator]
  }

  return items
})

const creatorsWithLimit = computed(() => {
  if (tooManyCreators.value) {
    return creators.value.slice(0, CREATOR_SHOW_LIMIT)
  }
  return creators.value.slice(1)
})

const verifyStatus = computed<VerifyStatus>(() => {
  const issued = useIssuedAssetsGraph().validateAssets(creators, assetAddress)
  const array = Array.from(issued.value?.values() || [])
  const hasSome = array?.some(info => info)
  if (!hasSome) {
    return 'unverified'
  }

  if (array.every(Boolean)) {
    return 'verified'
  }

  return 'partial'
})

const tooManyCreators = computed(
  () => creators.value.length > CREATOR_SHOW_LIMIT
)

const isLoaded = computed(() => asset.value && !asset.value.isLoading)
</script>

<template>
  <div v-if="isLoaded">
    <!--no creators at all including owner, might be that its EOA or not indexed -->
    <div
      v-if="(creators || []).length === 0"
      class="grid grid-cols-[max-content,auto] gap-1"
    ></div>
    <div v-else class="grid animate-fade-in grid-cols-[max-content,auto]">
      <div class="flex space-x-[-14px]">
        <NftListCardCreatorsProfileGraph
          v-for="(creator, index) in creatorsWithLimit || []"
          :creator="creator"
          :key="index"
          class="relative"
        />
        <NftListCardCreatorsProfileGraph
          v-if="creators[0]"
          :creator="creators[0]"
          :count="tooManyCreators ? creators.length - CREATOR_SHOW_LIMIT : 0"
          class="relative"
          has-name
        />
      </div>
      <div class="flex items-center justify-end">
        <lukso-tooltip
          v-if="verifyStatus === 'unverified'"
          variant="danger"
          :text="$formatMessage('asset_all_creators_unverified')"
          class="ml-2"
        >
          <lukso-icon
            name="cross-filled"
            color="red-55"
            secondary-color="neutral-100"
            size="small"
          ></lukso-icon>
        </lukso-tooltip>
        <lukso-tooltip
          v-if="verifyStatus === 'partial'"
          variant="danger"
          :text="$formatMessage('asset_all_creators_partial')"
          class="ml-2"
        >
          <lukso-icon
            name="cross-filled"
            color="red-55"
            secondary-color="neutral-100"
            size="small"
          ></lukso-icon>
        </lukso-tooltip>
        <lukso-tooltip
          v-if="verifyStatus === 'verified'"
          variant="success"
          :text="$formatMessage('asset_all_creators_verified')"
          class="ml-2"
        >
          <lukso-icon
            name="complete-filled"
            color="green-54"
            secondary-color="neutral-100"
            size="small"
          ></lukso-icon>
        </lukso-tooltip>
      </div>
    </div>
  </div>
  <div v-else class="mt-4 grid grid-cols-[max-content,auto] gap-1">
    <AppPlaceholderCircle class="size-6" />
    <div class="grid w-full flex-col gap-1">
      <AppPlaceholderLine class="w-1/3" />
      <AppPlaceholderLine class="w-1/2" />
    </div>
  </div>
</template>
