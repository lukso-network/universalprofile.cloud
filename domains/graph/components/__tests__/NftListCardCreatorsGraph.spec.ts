import { renderSuspended } from '@nuxt/test-utils/runtime'
import { beforeEach, describe, expect, it } from 'vitest'

import NftListCardCreatorsGraph from '../NftListCardCreatorsGraph.vue'

describe('NftListCardCreatorsGraph', () => {
  beforeEach(() => {
    useIntl().setupIntl(defaultConfig)
  })

  it('render loading state', async () => {
    const component = await renderSuspended(NftListCardCreatorsGraph)

    expect(component.html()).toMatchSnapshot()
  })

  it('render empty creators', async () => {
    const component = await renderSuspended(NftListCardCreatorsGraph, {
      props: {
        asset: {} as Asset,
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  it('render more creators then limit', async () => {
    const component = await renderSuspended(NftListCardCreatorsGraph, {
      props: {
        asset: {
          tokenCreatorsData: [
            {
              address: '0x123',
              name: 'John Doe',
              profileImage: [
                {
                  url: 'https://example.com/profile1-image.jpg',
                },
              ],
            } as Creator,
            {
              address: '0x456',
              name: 'Alice Doe',
              profileImage: [
                {
                  url: 'https://example.com/profile2-image.jpg',
                },
              ],
            } as Creator,
            {
              address: '0x789',
              name: 'Bob Doe',
              profileImage: [
                {
                  url: 'https://example.com/profile3-image.jpg',
                },
              ],
            } as Creator,
            {
              address: '0xabc',
              name: 'Charlie Doe',
              profileImage: [
                {
                  url: 'https://example.com/profile4-image.jpg',
                },
              ],
            } as Creator,
            {
              address: '0xdef',
              name: 'David Doe',
              profileImage: [
                {
                  url: 'https://example.com/profile5-image.jpg',
                },
              ],
            } as Creator,
          ],
        } as Asset,
      },
    })

    expect(component.html()).toMatchSnapshot()
  })

  describe('verification', () => {
    it('render all verified', async () => {
      const component = await renderSuspended(NftListCardCreatorsGraph, {
        props: {
          asset: {
            address: '0xabc',
            tokenCreatorsData: [
              {
                address: '0x123',
                name: 'John Doe',
                profileImage: [
                  {
                    url: 'https://example.com/profile1-image.jpg',
                  },
                ],
                issuedAssets: ['0xabc'],
              } as Creator,
              {
                address: '0x456',
                name: 'Alice Doe',
                profileImage: [
                  {
                    url: 'https://example.com/profile2-image.jpg',
                  },
                ],
                issuedAssets: ['0xabc'],
              } as Creator,
            ],
          } as Asset,
          hasVerification: true,
        },
      })

      expect(component.html()).toMatchSnapshot()
    })

    it('render some verified', async () => {
      const component = await renderSuspended(NftListCardCreatorsGraph, {
        props: {
          asset: {
            address: '0xabc',
            tokenCreatorsData: [
              {
                address: '0x123',
                name: 'John Doe',
                profileImage: [
                  {
                    url: 'https://example.com/profile1-image.jpg',
                  },
                ],
                issuedAssets: ['0xabc'],
              } as Creator,
              {
                address: '0x456',
                name: 'Alice Doe',
                profileImage: [
                  {
                    url: 'https://example.com/profile2-image.jpg',
                  },
                ],
                issuedAssets: ['0xcaf'],
              } as Creator,
            ],
          } as Asset,
          hasVerification: true,
        },
      })

      expect(component.html()).toMatchSnapshot()
    })

    it('render none verified', async () => {
      const component = await renderSuspended(NftListCardCreatorsGraph, {
        props: {
          asset: {
            address: '0xabc',
            tokenCreatorsData: [
              {
                address: '0x123',
                name: 'John Doe',
                profileImage: [
                  {
                    url: 'https://example.com/profile1-image.jpg',
                  },
                ],
                issuedAssets: ['0xcaf'],
              } as Creator,
              {
                address: '0x456',
                name: 'Alice Doe',
                profileImage: [
                  {
                    url: 'https://example.com/profile2-image.jpg',
                  },
                ],
                issuedAssets: ['0xcaf'],
              } as Creator,
            ],
          } as Asset,
          hasVerification: true,
        },
      })

      expect(component.html()).toMatchSnapshot()
    })

    it('render without verification', async () => {
      const component = await renderSuspended(NftListCardCreatorsGraph, {
        props: {
          asset: {
            tokenCreatorsData: [
              {
                address: '0x123',
                name: 'John Doe',
                profileImage: [
                  {
                    url: 'https://example.com/profile1-image.jpg',
                  },
                ],
              } as Creator,
            ],
          } as Asset,
          hasVerification: false,
        },
      })

      expect(component.html()).toMatchSnapshot()
    })
  })

  describe('large size', () => {
    it('render loading state', async () => {
      const component = await renderSuspended(NftListCardCreatorsGraph, {
        props: {
          isSmall: false,
        },
      })

      expect(component.html()).toMatchSnapshot()
    })

    it('render creators', async () => {
      const component = await renderSuspended(NftListCardCreatorsGraph, {
        props: {
          asset: {
            address: '0xabc',
            tokenCreatorsData: [
              {
                address: '0x123',
                name: 'John Doe',
                profileImage: [
                  {
                    url: 'https://example.com/profile1-image.jpg',
                  },
                ],
                issuedAssets: ['0xabc'],
              } as Creator,
            ],
          } as Asset,
          isSmall: false,
        },
      })

      expect(component.html()).toMatchSnapshot()
    })
  })
})
