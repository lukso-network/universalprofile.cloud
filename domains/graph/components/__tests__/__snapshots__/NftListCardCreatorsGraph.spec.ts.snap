// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NftListCardCreatorsGraph > large size > render creators 1`] = `
"<div id="test-wrapper">
  <div>
    <!--no creators at all including owner, might be that its EOA or not indexed -->
    <div class="grid animate-fade-in grid-cols-[max-content,auto]">
      <div class="flex space-x-[-26px]">
        <div class="flex h-10 relative">
          <lukso-profile size="small" profile-url="https://example.com/profile1-image.jpg" profile-address="0x123" has-identicon=""></lukso-profile>
          <div class="flex flex-col justify-center gap-1 pl-2">
            <div class="paragraph-inter-10-semi-bold text-neutral-60">Created by</div>
            <lukso-username name="John Doe" address="0x123" size="x-small" class="flex" name-color="neutral-20"></lukso-username>
          </div>
        </div>
      </div>
      <div class="flex items-center justify-end">
        <!--v-if-->
        <!--v-if-->
        <lukso-tooltip variant="success" text="Verified creators" class="ml-2">
          <lukso-icon name="complete-filled" color="green-54" secondary-color="neutral-100" size="small"></lukso-icon>
        </lukso-tooltip>
      </div>
    </div>
  </div>
</div>"
`;

exports[`NftListCardCreatorsGraph > large size > render loading state 1`] = `
"<div id="test-wrapper">
  <div class="grid grid-cols-[max-content,auto] gap-1">
    <div class="animate-pulse rounded-full bg-neutral-90 size-10"></div>
    <div class="grid w-full flex-col items-center gap-1 pl-2">
      <div class="animate-pulse rounded-4 bg-neutral-90 h-3 w-[60px]"></div>
      <div class="animate-pulse rounded-4 bg-neutral-90 h-3 w-[120px]"></div>
    </div>
  </div>
</div>"
`;

exports[`NftListCardCreatorsGraph > render empty creators 1`] = `
"<div id="test-wrapper">
  <div>
    <!--no creators at all including owner, might be that its EOA or not indexed -->
    <div class="grid grid-cols-[max-content,auto] gap-1"></div>
  </div>
</div>"
`;

exports[`NftListCardCreatorsGraph > render loading state 1`] = `
"<div id="test-wrapper">
  <div class="grid grid-cols-[max-content,auto] gap-1">
    <div class="animate-pulse rounded-full bg-neutral-90 size-6"></div>
    <div class="grid w-full flex-col items-center gap-1 pl-1">
      <div class="animate-pulse rounded-4 bg-neutral-90 h-3 w-[40px]"></div>
      <div class="animate-pulse rounded-4 bg-neutral-90 h-3 w-[80px]"></div>
    </div>
  </div>
</div>"
`;

exports[`NftListCardCreatorsGraph > render more creators then limit 1`] = `
"<div id="test-wrapper">
  <div>
    <!--no creators at all including owner, might be that its EOA or not indexed -->
    <div class="grid animate-fade-in grid-cols-[max-content,auto]">
      <div class="flex space-x-[-14px]">
        <div class="flex h-6 relative">
          <lukso-profile size="x-small" profile-url="https://example.com/profile1-image.jpg" profile-address="0x123" has-identicon=""></lukso-profile>
          <!--v-if-->
        </div>
        <div class="flex h-6 relative">
          <lukso-profile size="x-small" profile-url="https://example.com/profile2-image.jpg" profile-address="0x456" has-identicon=""></lukso-profile>
          <!--v-if-->
        </div>
        <div class="flex h-6 relative">
          <lukso-profile size="x-small" profile-url="https://example.com/profile3-image.jpg" profile-address="0x789" has-identicon=""></lukso-profile>
          <!--v-if-->
        </div>
        <div class="flex h-6 relative">
          <lukso-profile size="x-small" profile-url="https://example.com/profile4-image.jpg" profile-address="0xabc" has-identicon=""></lukso-profile>
          <!--v-if-->
        </div>
        <div class="flex h-6 relative">
          <div class="flex items-center justify-center rounded-full bg-neutral-100 outline outline-2 outline-neutral-90 paragraph-inter-10-semi-bold size-6"> +1</div>
          <div class="flex flex-col justify-center pl-1">
            <div class="paragraph-inter-10-semi-bold text-neutral-60">Created by</div>
            <lukso-username name="John Doe" address="0x123" size="x-small" class="flex" name-color="neutral-20"></lukso-username>
          </div>
        </div>
      </div>
      <div class="flex items-center justify-end">
        <lukso-tooltip variant="danger" text="Un-verified creators" class="ml-2">
          <lukso-icon name="cross-filled" color="red-55" secondary-color="neutral-100" size="small"></lukso-icon>
        </lukso-tooltip>
        <!--v-if-->
        <!--v-if-->
      </div>
    </div>
  </div>
</div>"
`;

exports[`NftListCardCreatorsGraph > verification > render all verified 1`] = `
"<div id="test-wrapper">
  <div>
    <!--no creators at all including owner, might be that its EOA or not indexed -->
    <div class="grid animate-fade-in grid-cols-[max-content,auto]">
      <div class="flex space-x-[-14px]">
        <div class="flex h-6 relative">
          <lukso-profile size="x-small" profile-url="https://example.com/profile2-image.jpg" profile-address="0x456" has-identicon=""></lukso-profile>
          <!--v-if-->
        </div>
        <div class="flex h-6 relative">
          <lukso-profile size="x-small" profile-url="https://example.com/profile1-image.jpg" profile-address="0x123" has-identicon=""></lukso-profile>
          <div class="flex flex-col justify-center pl-1">
            <div class="paragraph-inter-10-semi-bold text-neutral-60">Created by</div>
            <lukso-username name="John Doe" address="0x123" size="x-small" class="flex" name-color="neutral-20"></lukso-username>
          </div>
        </div>
      </div>
      <div class="flex items-center justify-end">
        <!--v-if-->
        <!--v-if-->
        <lukso-tooltip variant="success" text="Verified creators" class="ml-2">
          <lukso-icon name="complete-filled" color="green-54" secondary-color="neutral-100" size="small"></lukso-icon>
        </lukso-tooltip>
      </div>
    </div>
  </div>
</div>"
`;

exports[`NftListCardCreatorsGraph > verification > render none verified 1`] = `
"<div id="test-wrapper">
  <div>
    <!--no creators at all including owner, might be that its EOA or not indexed -->
    <div class="grid animate-fade-in grid-cols-[max-content,auto]">
      <div class="flex space-x-[-14px]">
        <div class="flex h-6 relative">
          <lukso-profile size="x-small" profile-url="https://example.com/profile2-image.jpg" profile-address="0x456" has-identicon=""></lukso-profile>
          <!--v-if-->
        </div>
        <div class="flex h-6 relative">
          <lukso-profile size="x-small" profile-url="https://example.com/profile1-image.jpg" profile-address="0x123" has-identicon=""></lukso-profile>
          <div class="flex flex-col justify-center pl-1">
            <div class="paragraph-inter-10-semi-bold text-neutral-60">Created by</div>
            <lukso-username name="John Doe" address="0x123" size="x-small" class="flex" name-color="neutral-20"></lukso-username>
          </div>
        </div>
      </div>
      <div class="flex items-center justify-end">
        <lukso-tooltip variant="danger" text="Un-verified creators" class="ml-2">
          <lukso-icon name="cross-filled" color="red-55" secondary-color="neutral-100" size="small"></lukso-icon>
        </lukso-tooltip>
        <!--v-if-->
        <!--v-if-->
      </div>
    </div>
  </div>
</div>"
`;

exports[`NftListCardCreatorsGraph > verification > render some verified 1`] = `
"<div id="test-wrapper">
  <div>
    <!--no creators at all including owner, might be that its EOA or not indexed -->
    <div class="grid animate-fade-in grid-cols-[max-content,auto]">
      <div class="flex space-x-[-14px]">
        <div class="flex h-6 relative">
          <lukso-profile size="x-small" profile-url="https://example.com/profile2-image.jpg" profile-address="0x456" has-identicon=""></lukso-profile>
          <!--v-if-->
        </div>
        <div class="flex h-6 relative">
          <lukso-profile size="x-small" profile-url="https://example.com/profile1-image.jpg" profile-address="0x123" has-identicon=""></lukso-profile>
          <div class="flex flex-col justify-center pl-1">
            <div class="paragraph-inter-10-semi-bold text-neutral-60">Created by</div>
            <lukso-username name="John Doe" address="0x123" size="x-small" class="flex" name-color="neutral-20"></lukso-username>
          </div>
        </div>
      </div>
      <div class="flex items-center justify-end">
        <!--v-if-->
        <lukso-tooltip variant="danger" text="Not all creators are verified" class="ml-2">
          <lukso-icon name="cross-filled" color="red-55" secondary-color="neutral-100" size="small"></lukso-icon>
        </lukso-tooltip>
        <!--v-if-->
      </div>
    </div>
  </div>
</div>"
`;

exports[`NftListCardCreatorsGraph > verification > render without verification 1`] = `
"<div id="test-wrapper">
  <div>
    <!--no creators at all including owner, might be that its EOA or not indexed -->
    <div class="grid animate-fade-in grid-cols-[max-content,auto]">
      <div class="flex space-x-[-14px]">
        <div class="flex h-6 relative">
          <lukso-profile size="x-small" profile-url="https://example.com/profile1-image.jpg" profile-address="0x123" has-identicon=""></lukso-profile>
          <div class="flex flex-col justify-center pl-1">
            <div class="paragraph-inter-10-semi-bold text-neutral-60">Created by</div>
            <lukso-username name="John Doe" address="0x123" size="x-small" class="flex" name-color="neutral-20"></lukso-username>
          </div>
        </div>
      </div>
      <!--v-if-->
    </div>
  </div>
</div>"
`;
